backend/de.metas.handlingunits.base/src/main/java/org/adempiere/mm/attributes/spi/impl/HUVendorBPartnerAttributeValuesProvider.java

```diff
@@ -0,0 +1,303 @@
+		);
+
+		//
+		// Get from cache / Load
+		return hu2Vendors.getOrLoad(cacheKey, () -> retrieveVendorKeyNamePairs(currentVendorId));
+	}
+
+	@Nullable
+	private static String normalizeValueKey(final Object valueKey)
+	{
+		if (valueKey == null)
+		{
+			return null;
+		}
+		else if (valueKey instanceof Number)
+		{
+			final int valueKeyAsInt = ((Number)valueKey).intValue();
+			return String.valueOf(valueKeyAsInt);
+		}
+		else
+		{
+			return valueKey.toString();
+		}
+	}
+
+	@Override
+	@Nullable
+	public NamePair getAttributeValueOrNull(final Evaluatee evalCtx, final Object valueKey)
+	{
+		final List<? extends NamePair> availableValues = getAvailableValues(evalCtx);
+		if (availableValues.isEmpty())
+		{
+			return null;
+		}
+
+		// Normalize the valueKey
+		final String valueKeyNormalized = normalizeValueKey(valueKey);
+
+		for (final NamePair vnp : availableValues)
+		{
+			if (Objects.equals(vnp.getID(), valueKeyNormalized))
+			{
+				return vnp;
+			}
+		}
+
+		return null;
+	}
+
+	@Nullable
+	@Override
+	public AttributeValueId getAttributeValueIdOrNull(final Object valueKey)
+	{
+		return null;
+	}
+
+	/**
+	 * Retrieve allowed vendor KeyNamePairs.
+	 *
+	 * Mainly it will contain following values:
+	 * <ul>
+	 * <li>{@link #getNullValue()}
+	 * <li>available vendors
+	 * <li>current HU's vendor if not present in the list above
+	 * </ul>
+	 *
+	 * @param currentVendorBPartnerId current vendor ID
+	 * @return available values
+	 */
+	private static List<KeyNamePair> retrieveVendorKeyNamePairs(
+			final int currentVendorBPartnerId)
+	{
+		final Properties ctx = Env.getCtx();
+
+		final List<KeyNamePair> vendors = new ArrayList<>();
+		vendors.add(staticNullValue());
+
+		final List<KeyNamePair> vendorsCached = HUVendorBPartnerAttributeValuesProvider.vendors.getOrLoad(currentVendorBPartnerId, () -> retrieveVendorKeyNamePairs());
+			vendors.addAll(vendorsCached);
+
+		addVendor(vendors, ctx, currentVendorBPartnerId);
+
+		return vendors;
+	}
+
+	private static void addVendor(final List<KeyNamePair> vendors, final Properties ctx, final int bpartnerId)
+	{
+		if (bpartnerId <= 0)
+		{
+			return;
+		}
+
+		//
+		// Check if our BPartner is already in the list.
+		// If yes, there is no point to add it.
+		if (vendors != null)
+		{
+			for (final KeyNamePair vendor : vendors)
+			{
+				if (vendor == null)
+				{
+					continue;
+				}
+				if (vendor.getKey() == bpartnerId)
+				{
+					// our bpartner is already in the list
+					return;
+				}
+			}
+		}
+
+		//
+		// Load the new BPartner and add it to our list
+		final I_C_BPartner bpartner = InterfaceWrapperHelper.create(ctx, bpartnerId, I_C_BPartner.class, ITrx.TRXNAME_None);
+		if (bpartner == null)
+		{
+			return;
+		}
+		final KeyNamePair bpartnerKNP = toKeyNamePair(bpartner);
+		vendors.add(bpartnerKNP);
+	}
+
+	private static List<KeyNamePair> retrieveVendorKeyNamePairs()
+	{
+		return bPartnerDAO.retrieveVendors()
```

Comment:
there we get all active vendors..and i wonder, waht if there are 1M of those vendors in metasfresh.
i think that's very hypothetical..but just in case a better solution pops in your head..